microsoft/testsuites/cpu/common.py

```diff
@@ -0,0 +1,131 @@
+    if config_result.exit_code != 0:
+        raise SkippedException("This distro doesn't support cpu hot plug.")
+
+
+def set_vmbus_channels_target_cpu(node: Node) -> Dict[str, str]:
+    uname = node.tools[Uname]
+    kernel_version = uname.get_linux_information().kernel_version
+    dmesg = node.tools[Dmesg]
+    lsvmbus = node.tools[Lsvmbus]
+    vmbus_version = dmesg.get_vmbus_version()
+    file_path_list: Dict[str, str] = {}
+    if kernel_version >= "5.8.0" and vmbus_version >= "4.1.0":
+        # save the raw cpu number for each channel for restoring later.
+        channels = lsvmbus.get_device_channels_from_lsvmbus(force_run=True)
+        for channel in channels:
+            for channel_vp_map in channel.channel_vp_map:
+                target_cpu = channel_vp_map.target_cpu
+                if target_cpu == "0":
+                    continue
+                file_path_list[
+                    get_interrupts_assigned_to_cpu(
+                        channel.device_id, channel_vp_map.rel_id
+                    )
+                ] = target_cpu
+        # set all vmbus channel interrupts go into cpu 0.
+        set_cpu_interrupts_assigned_to(file_path_list, node, "0")
+        return file_path_list
+
+
+def get_idle_cpu_set(node: Node) -> List[str]:
+    lsvmbus = node.tools[Lsvmbus]
+    channels = lsvmbus.get_device_channels_from_lsvmbus(force_run=True)
+    cpu_in_used = set()
+    for channel in channels:
+        for channel_vp_map in channel.channel_vp_map:
+            target_cpu = channel_vp_map.target_cpu
+            if target_cpu == "0":
+                continue
+            cpu_in_used.add(target_cpu)
+
+    cpu_count = node.tools[Lscpu].get_core_count()
+    all_cpu = list(range(1, cpu_count))
+    idle_cpu = [str(x) for x in all_cpu if str(x) not in cpu_in_used]
+    return idle_cpu
+
+
+def set_idle_cpu_offline_online(node: Node, idle_cpu: List[str]) -> None:
+    for target_cpu in idle_cpu:
+        result = set_cpu_state(target_cpu, CPUState.OFFLINE, node)
+        if result:
+            # bring cpu back to its original state
+            reset = set_cpu_state(target_cpu, CPUState.ONLINE, node)
+            exception_message = (
+                f"expected cpu{target_cpu} state: {CPUState.ONLINE}(online), "
+                f"actual state: {CPUState.OFFLINE}(offline)."
+            )
+            if not reset:
+                raise BadEnvironmentStateException(
+                    exception_message,
+                    f"the test failed leaving cpu{target_cpu} in a bad state.",
+                )
+
+
+def verify_cpu_hot_plug(node: Node, run_times: int = 1) -> None:
+    try:
+        skip_test(node)
+        for _ in range(run_times):
+            file_path_list = set_vmbus_channels_target_cpu(node)
+            idle_cpu = get_idle_cpu_set(node)
+            if 0 == len(idle_cpu):
+                raise SkippedException("no idle cpu can be set offline or online.")
```

Comment:
add reason, why no idle cpu, or how the idle is defined.

Reply:
added.